File: apps/mobile/modules/task-handoff-browser/android/src/main/java/expo/modules/taskhandoffbrowser/BrowserTunnelProtocol.kt
package expo.modules.taskhandoffbrowser

import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.ByteOrder

internal object BrowserTunnelProtocol {
  const val VERSION = "2026-08-29"
  const val HEADER_BYTES = 9
  const val INITIAL_WINDOW = 256 * 1024
  const val MAX_DATA_BYTES = 64 * 1024
  const val MAX_CONTROL_BYTES = 4 * 1024
  const val MAX_WINDOW = 16 * 1024 * 1024

  enum class FrameType(val value: Byte) {
    OPEN(1), OPEN_OK(2), DATA(3), HALF_CLOSE(4), CLOSE(5), ERROR(6), WINDOW_UPDATE(7);
    companion object { fun from(value: Byte) = entries.firstOrNull { it.value == value } }
  }

  data class Frame(val type: FrameType, val streamId: Long, val payload: ByteArray)

  fun hello() = JSONObject()
    .put("type", "browser-tunnel.hello")
    .put("protocolVersion", VERSION)
    .put("initialWindowBytes", INITIAL_WINDOW)
    .toString()

  fun open(streamId: Long, host: String, port: Int): ByteArray {
    require(host.isNotBlank() && host.toByteArray().size <= 255 && !host.any { it.isWhitespace() || it in "/?#@\\[]" })
    require(port in 1..65535)
    return encode(Frame(FrameType.OPEN, streamId, JSONObject().put("host", host).put("port", port).toString().toByteArray()))
  }

  fun encode(frame: Frame): ByteArray {
    require(frame.streamId in 1..0xffffffffL)
    val max = if (frame.type == FrameType.DATA) MAX_DATA_BYTES else MAX_CONTROL_BYTES
    require(frame.payload.size <= max)
    require(frame.type !in setOf(FrameType.OPEN_OK, FrameType.HALF_CLOSE, FrameType.CLOSE) || frame.payload.isEmpty())
    return ByteBuffer.allocate(HEADER_BYTES + frame.payload.size).order(ByteOrder.BIG_ENDIAN)
      .put(frame.type.value).putInt(frame.streamId.toInt()).putInt(frame.payload.size).put(frame.payload).array()
  }

  fun decode(input: ByteArray): Frame {
    require(input.size >= HEADER_BYTES)
    val buffer = ByteBuffer.wrap(input).order(ByteOrder.BIG_ENDIAN)
    val type = requireNotNull(FrameType.from(buffer.get()))
    val streamId = buffer.int.toLong() and 0xffffffffL
    val length = buffer.int
    require(streamId > 0 && length == buffer.remaining())
    val payload = ByteArray(length)
    buffer.get(payload)
    return Frame(type, streamId, payload).also { encode(it) }
  }

  fun windowUpdate(streamId: Long, bytes: Int): ByteArray {
    require(bytes in 1..MAX_WINDOW)
    return encode(Frame(FrameType.WINDOW_UPDATE, streamId, ByteBuffer.allocate(4).putInt(bytes).array()))
  }

  fun windowBytes(frame: Frame): Int {
    require(frame.type == FrameType.WINDOW_UPDATE && frame.payload.size == 4)
    return ByteBuffer.wrap(frame.payload).int.also { require(it in 1..MAX_WINDOW) }
  }
}

File: apps/mobile/modules/task-handoff-browser/ios/BrowserTunnelProtocol.swift
import Foundation

internal enum BrowserTunnelProtocol {
  static let version = "2026-08-29"
  static let headerBytes = 9
  static let initialWindow = 256 * 1024
  static let maxDataBytes = 64 * 1024
  static let maxControlBytes = 4 * 1024
  static let maxWindow = 16 * 1024 * 1024

  enum FrameType: UInt8 {
    case open = 1
    case openOk = 2
    case data = 3
    case halfClose = 4
    case close = 5
    case error = 6
    case windowUpdate = 7
  }

  struct Frame {
    let type: FrameType
    let streamId: UInt32
    let payload: Data
  }

  struct Ready: Decodable {
    let type: String
    let protocolVersion: String
    let initialWindowBytes: Int
  }

  struct TunnelError: Decodable {
    let code: String
    let message: String
  }

  static func hello() throws -> String {
    let value: [String: Any] = [
      "type": "browser-tunnel.hello",
      "protocolVersion": version,
      "initialWindowBytes": initialWindow,
    ]
    return String(decoding: try JSONSerialization.data(withJSONObject: value), as: UTF8.self)
  }

  static func open(streamId: UInt32, host: String, port: UInt16) throws -> Data {
    guard !host.isEmpty, host.utf8.count <= 255,
          host.rangeOfCharacter(from: CharacterSet.whitespacesAndNewlines.union(CharacterSet(charactersIn: "/?#@\\[]"))) == nil else {
      throw BrowserTunnelProtocolError("Browser tunnel target is invalid.")
    }
    return try encode(Frame(
      type: .open,
      streamId: streamId,
      payload: JSONSerialization.data(withJSONObject: ["host": host, "port": Int(port)])
    ))
  }

  static func windowUpdate(streamId: UInt32, bytes: Int) throws -> Data {
    guard bytes > 0, bytes <= maxWindow else { throw BrowserTunnelProtocolError("Browser tunnel window update is invalid.") }
    var value = UInt32(bytes).bigEndian
    return try encode(Frame(type: .windowUpdate, streamId: streamId, payload: Data(bytes: &value, count: 4)))
  }

  static func encode(_ frame: Frame) throws -> Data {
    guard frame.streamId > 0 else { throw BrowserTunnelProtocolError("Browser tunnel stream id must be positive.") }
    let maxPayload = frame.type == .data ? maxDataBytes : maxControlBytes
    guard frame.payload.count <= maxPayload else { throw BrowserTunnelProtocolError("Browser tunnel frame is too large.") }
    if [.openOk, .halfClose, .close].contains(frame.type), !frame.payload.isEmpty {
      throw BrowserTunnelProtocolError("Browser tunnel control payload must be empty.")
    }
    var output = Data(capacity: headerBytes + frame.payload.count)
    output.append(frame.type.rawValue)
    var streamId = frame.streamId.bigEndian
    var length = UInt32(frame.payload.count).bigEndian
    output.append(Data(bytes: &streamId, count: 4))
    output.append(Data(bytes: &length, count: 4))
    output.append(frame.payload)
    return output
  }

  static func decode(_ input: Data) throws -> Frame {
    guard input.count >= headerBytes,
          let type = FrameType(rawValue: input[input.startIndex]),
          let streamId = uint32(input, offset: 1),
          let length = uint32(input, offset: 5),
          streamId > 0,
          Int(length) == input.count - headerBytes else {
      throw BrowserTunnelProtocolError("Browser tunnel frame is invalid.")
    }
    let payload = input.subdata(in: headerBytes..<input.count)
    let frame = Frame(type: type, streamId: streamId, payload: payload)
    _ = try encode(frame)
    return frame
  }

  static func windowBytes(_ frame: Frame) throws -> Int {
    guard frame.type == .windowUpdate, frame.payload.count == 4,
          let bytes = uint32(frame.payload, offset: 0), bytes > 0, bytes <= maxWindow else {
      throw BrowserTunnelProtocolError("Browser tunnel window update is invalid.")
    }
    return Int(bytes)
  }

  private static func uint32(_ data: Data, offset: Int) -> UInt32? {
    guard offset >= 0, offset + 4 <= data.count else { return nil }
    return data[offset..<(offset + 4)].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
  }
}

internal struct BrowserTunnelProtocolError: Error {
  let message: String
  init(_ message: String) { self.message = message }
}
